NEIGHBORHOODS IN THE CITY OF TORONTO

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


First import some libraries

Download and explore data set

First import urllib and beautiful soup in order to explore data from wikipedia

In [2]:
# import libraries
import urllib.request
#import urllib2
#from bs4 import BeautifulSoup

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = urllib.request.urlopen(url)

In [4]:
from bs4 import BeautifulSoup

In [5]:
soup = BeautifulSoup(page, "lxml")

In [6]:
#print(soup.prettify())
#soup.title

We can check the HTML code from the web page with 'prettify' and
We can verify the title of the web page

In [7]:
all_tables=soup.find_all("table")
#all_tables

 **Now let's find the tables in the wikipedia web page**

In [8]:
right_table=soup.find('table', class_='wikitable sortable')
#right_table

**roll dow until find 'wikitable sortable' it is the table that we need**

In [9]:
A=[]
B=[]
C=[]
#D=[]
#E=[]
 
for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))
        

**Time to extract the content string from within each part of table marked as < td > and add them to the lists from A to C using the -find- option**

In [10]:
df=pd.DataFrame(A,columns=['Postal Code'])
df['Borough']=B
df['Neighbourhood']=C
df.head(15)



Postal Code           Borough     Neighbourhood
0          M1A      Not assigned     Not assigned

1          M2A      Not assigned     Not assigned

2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park     Not assigned

9          M8A      Not assigned     Not assigned

10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern
13         M2B      Not assigned     Not assigned

14         M3B        North York  Don Mills North

In [11]:
df.drop(df.loc[df['Borough']=='Not assigned'].index, inplace=True)
df.head(15)

Postal Code           Borough     Neighbourhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park     Not assigned

10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern
14         M3B        North York  Don Mills North

15         M4B         East York  Woodbine Gardens
16         M4B         East York     Parkview Hill
17         M5B  Downtown Toronto          Ryerson

18         M5B  Downtown Toronto  Garden District

**Now cleaning the data frame removing 'Not assigned' values**

In [12]:
df1 =df.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(lambda x: ', '.join(x.astype(str))).reset_index()
df1.head(15)

Postal Code      Borough                                      Neighbourhood
0          M1B  Scarborough                                     Rouge, Malvern
1          M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2          M1E  Scarborough                Guildwood\n, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                        Cedarbrae\n
5          M1J  Scarborough                                Scarborough Village
6          M1K  Scarborough      East Birchmount Park\n, Ionview, Kennedy Park
7          M1L  Scarborough                    Clairlea, Golden Mile, Oakridge
8          M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West\n
9          M1N  Scarborough                      Birch Cliff, Cliffside West\n
10         M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...
11         M1R  Scarborough                                  Maryvale, Wexford
12         M1S  Scarborough                                          Agincourt
13         M1T  Scarborough        Clarks Corners\n, Sullivan\n, Tam O'Shanter
14         M1V  Scarborough  Agincourt North, L'Amoreaux East\n, Milliken, ...

**Using groupby to put the same Neighbourhood associated to a postal code**

In [15]:
df1['Neighbourhood'].replace(['Not assigned\n'], 'Queens Park',inplace=True)
df1.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

In [64]:
# I WANT TO SORT THE DATA FRAME FOR ORGANIZE THE DATA
sort_by_code = df1.sort_values('Postal Code')
sort_by_code.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

**The data frame is sorted just for organize the data**

**for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park. For me, after I sorted the data frame the cell is the 85th**

In [65]:
sort_by_code.shape

(103, 3)

**103 ROWS**

**This is the same data frame with another name. **

In [66]:
sortdf1 = df1.sort_values('Postal Code')
sortdf1.head()


Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough     Guildwood\n, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                             Cedarbrae\n

**I decide to use de Geospatial_Coordinates.csv after spend a lot of time trying with geocode. The results were none all the time, or falling in errors
**

**Faster in this way**

In [67]:
dft = pd.read_csv('Geospatial_Coordinates.csv')
dft.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

**Just delete the column Postal code and keep the latitude and longitude columns**

In [59]:
del dft['Postal Code']
dft.head()

Latitude  Longitude
0  43.806686 -79.194353
1  43.784535 -79.160497
2  43.763573 -79.188711
3  43.770992 -79.216917
4  43.773136 -79.239476

**Finally concatenate the two data frames**

In [60]:
dfTN=pd.concat([sortdf1,dft], axis=1)


In [62]:
dfTN.head(15)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   

                                        Neighbourhood   Latitude  Longitude  
0                                      Rouge, Malvern  43.806686 -79.194353  
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                 Guildwood\n, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                         Cedarbrae\n  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6       East Birchmount Park\n, Ionview, Kennedy Park  43.727929 -79.262029  
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8   Cliffcrest, Cliffside, Scarborough Village West\n  43.716316 -79.239476  
9                       Birch Cliff, Cliffside West\n  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13        Clarks Corners\n, Sullivan\n, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East\n, Milliken, ...  43.815252 -79.284577